public/app/features/plugins/extensions/logs/filterTransformation.ts

```diff
@@ -0,0 +1,123 @@
+            return current;
+          }
+
+          for (const frame of current) {
+            const pluginIdIndex = frame.fields.findIndex((f) => f.name === 'pluginId');
+            const extensionPointIdIndex = frame.fields.findIndex((f) => f.name === 'extensionPointId');
+            const severityIndex = frame.fields.findIndex((f) => f.name === 'severity');
+
+            if (pluginIdIndex === -1 && !isEmpty(filter.pluginIds)) {
+              continue;
+            }
+
+            if (extensionPointIdIndex === -1 && !isEmpty(filter.extensionPointIds)) {
+              continue;
+            }
+
+            if (severityIndex === -1 && !isEmpty(filter.severity)) {
+              continue;
+            }
+
+            const target: PartialDataFrame = {
+              ...frame,
+              fields: frame.fields.map((f) => ({
+                ...f,
+                values: [],
+              })),
+            };
+
+            for (let index = 0; index < frame.length; index++) {
+              const pluginId = frame.fields[pluginIdIndex].values[index];
+              const extensionPointId = frame.fields[extensionPointIdIndex].values[index];
+              const severity = frame.fields[severityIndex].values[index];
+
+              if (!isEmpty(filter.pluginIds) && filter.pluginIds?.has(pluginId)) {
+                if (!isEmpty(filter.extensionPointIds)) {
+                  if (!filter.extensionPointIds?.has(extensionPointId)) {
+                    continue;
+                  }
+                }
+
+                if (!isEmpty(filter.severity)) {
+                  if (!filter.severity?.has(severity)) {
+                    continue;
+                  }
+                }
+
+                copyRow(frame, target, index);
+                continue;
+              }
+
+              if (!isEmpty(filter.extensionPointIds) && filter.extensionPointIds?.has(extensionPointId)) {
+                if (!isEmpty(filter.pluginIds)) {
+                  if (!filter.pluginIds?.has(pluginId)) {
+                    continue;
+                  }
+                }
+
+                if (!isEmpty(filter.severity)) {
+                  if (!filter.severity?.has(severity)) {
+                    continue;
+                  }
+                }
+
+                copyRow(frame, target, index);
+                continue;
+              }
+
+              if (!isEmpty(filter.severity) && filter.severity?.has(severity)) {
+                if (!isEmpty(filter.pluginIds)) {
+                  if (!filter.pluginIds?.has(pluginId)) {
+                    continue;
+                  }
+                }
+
+                if (!isEmpty(filter.extensionPointIds)) {
+                  if (!filter.extensionPointIds?.has(extensionPointId)) {
+                    continue;
+                  }
+                }
+
+                copyRow(frame, target, index);
+                continue;
```

Comment:
Hmm, could we reduce the nesting like this? (If I understand correctly we filter out a row if any of the filters don't match)

```ts
// Plugin id doesn't match
if (
  !isEmpty(filter.pluginIds) && 
  !filter.pluginIds?.has(pluginId)
) {
  continue;
}

// Extension point id doesn't match
if (
  !isEmpty(filter.extensionPointIds) && 
  !filter.extensionPointIds?.has(extensionPointId)
) {
  continue;
}

// Severity doesn't match
if (
  !isEmpty(filter.severity) && 
  !filter.severity?.has(severity)
) {
  continue;
}

copyRow(frame, target, index);
filtered.push(createDataFrame(target));
```

Reply:
🎉